server/frida_server.go

```diff
@@ -0,0 +1,533 @@
+//
+// ```
+// {
+//   "batch": [
+//     { "deviceId": "someid", "payload": "encryptedBlob" }
+//   ]
+// }
+// ```
+//
+// Returns whatever
+func (server *Server) postMessage(rw http.ResponseWriter, req *http.Request) {
+	authHeader := req.Header.Get("Authorization")
+	if authHeader == "" || len(authHeader) < 8 {
+		http.Error(rw, "Not authorized", http.StatusNotFound)
+		return
+	}
+	senderDeviceId := strings.TrimSpace(authHeader[7:])
+
+	type Batch struct {
+		Batch []*IncomingMessage `json:"batch"`
+	}
+	var msgs Batch
+	e := json.NewDecoder(req.Body).Decode(&msgs)
+	if e != nil {
+		http.Error(rw, fmt.Sprintf("%s", e), http.StatusInternalServerError)
+		return
+	}
+
+	server.MessageStorage.lock.Lock()
+	batch := server.MessageStorage.db.NewIndexedBatch()
+	var seqID uint64 = 0
+	var newSeqCount []byte
+	seqCount, closer, e := batch.Get([]byte{0})
+	if e == nil {
+		seqID = binary.LittleEndian.Uint64(seqCount)
+		closer.Close()
+		newSeqCount = make([]byte, 8)
+		binary.LittleEndian.PutUint64(newSeqCount, seqID + 1)
+	} else {
+		seqCount = make([]byte, 8)
+		binary.LittleEndian.PutUint64(seqCount, seqID)
+
+		newSeqCount = make([]byte, 8)
+		binary.LittleEndian.PutUint64(newSeqCount, seqID + 1)
+	}
+	batch.Set([]byte{0}, newSeqCount, pebble.Sync)
+	var tmsgs []*Message
+	for _, msg := range msgs.Batch {
+		tmsg := Message{}
+
+		tmsg.To = msg.DeviceId
+		tmsg.Outgoing.Payload = msg.Payload
+		tmsg.Outgoing.Sender = senderDeviceId
+		tmsg.Outgoing.SeqID = seqID
+
+		tmsgs = append(tmsgs, &tmsg)
+
+		k := append(append([]byte(msg.DeviceId), 0), seqCount...)
+		msgStorage, _ := json.Marshal(&tmsg.Outgoing)
+		batch.Set(k, msgStorage, pebble.Sync)
+	}
+	batch.Commit(pebble.Sync)
+	event := MessageEvent{Messages: tmsgs, SeqID: seqID}
+
+	server.Notifier <- &event
+	server.MessageStorage.lock.Unlock()
+	rw.Write([]byte("{}"))
+}
+
+func (server *Server) serveEvents(rw http.ResponseWriter, req *http.Request) {
+
+	// Make sure that the writer supports flushing.
+	//
+	flusher, ok := rw.(http.Flusher)
+
+	if !ok {
+		http.Error(rw, "Streaming unsupported!", http.StatusInternalServerError)
+		return
+	}
+
+	rw.Header().Set("Content-Type", "text/event-stream")
+	rw.Header().Set("Cache-Control", "no-cache")
+	rw.Header().Set("Connection", "keep-alive")
+	rw.Header().Set("Access-Control-Allow-Origin", "*")
+
+	authHeader := req.Header.Get("Authorization")
+	if authHeader == "" || len(authHeader) < 8 {
+		http.Error(rw, "Not authorized", http.StatusUnauthorized)
+		return
+	}
+	deviceId := strings.TrimSpace(authHeader[7:])
+
+	messageChan := ClientChan{
+		DeviceId: deviceId,
+		Channel:  make(chan Notification, 10),
+	}
+
+	server.newClients <- messageChan
+
+	// Remove this client from the map of connected clients
+	// when this handler exits.
+	defer func() {
+		server.closingClients <- deviceId
+	}()
+
+	notify := req.Context().Done()
+
+	go func() {
+		<-notify
+		server.closingClients <- deviceId
+	}()
```

Comment:
What is this req.Context().Done() stuff for?

Reply:
`Context().Done()` is a channel that yields when the "context" closes, in other words, in this case, when the client disconnects.